TelestionTeam/telestion-client
# REBASE: feat/common/loading-indicator improves TypeScript support and fixes inferred type error

## BREAKING CHANGES

- The useDependencyTimeout now returns `true` if all dependencies are defined and `false` if not all dependencies are defined yet. Before the return value was inverted. (9ca4b6a)
- The loading indicator gives the children function the current dependencies which are always defined. (e482423)

## Features

- The loading indicator now supports variadic tuples for the dependency list via generics and gives the children function the current dependencies which are always defined. (e482423)
  Now the LoadingIndicator is simpler to use, for example:
  ```tsx
  const [position, setPosition] = useState<Position>();
  return (
    <LoadingIndicator dependencies={[position]}>
      // "currentPos" is "position" from above but always defined
      {(currentPos) => (
        <p>{currentPos}</p>
      )}
    </LoadingIndicator>
  );
  ```
- Make useDependencyTimeout generic and usable as type guard for the dependency list. (9ca4b6a)
  It makes the hook useful as a condition in if-statements, for example:
  ```tsx
  const [position, setPosition] = useState<Position>();
  
  // throws if no position received after 5 seconds
  if (useDependencyTimeout(5000, [position])) {
    // type guarded - "position" is always defined
    return <p>Latest position: {position}</p>;
  }
  
  return <p>Waiting for incoming data</p>;
  ```
- Add utility type to make all elements in a tuple undefinable. (419970f)
- Add utility function to check if a value is defined. (51a64d3)

## Fixes

- It fixes a type error where TypeScript wrongly infers the return type of the loading indicator component. (e482423)
